querySelector() throws exception for -webkit-prefixed pseudo elements

annevk opened the pull request:
#### a23b77e8bca1e6a510f6b23e688695ac68ccdb98
<pre>
querySelector() throws exception for -webkit-prefixed pseudo elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=149160">https://bugs.webkit.org/show_bug.cgi?id=149160</a>
<a href="https://rdar.apple.com/99299129">rdar://99299129</a>

Reviewed by Antti Koivisto.

SelectorHasInvalidSelectorFunctor which is only used by querySelector()
and friends treated custom pseudo-elements as invalid. This is wrong.
querySelector(&quot;::placeholder&quot;) and querySelector(&quot;::-webkit-asdf&quot;) are
perfectly acceptable inputs, they are just supposed to return null.

Also do some minor code cleanup while in the general area.

Canonical link: <a href="https://commits.webkit.org/272337@main">https://commits.webkit.org/272337@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de86f5a4212a4b741fbd5b28103f3c63148fafae

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10031 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33856 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28428 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32129 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12376 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7281 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28064 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31698 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8447 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28000 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7266 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7424 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27904 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35197 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28511 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28355 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33561 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7498 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5528 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31401 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9161 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7364 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8191 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8009 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -175,7 +175,7 @@ class SelectorHasInvalidSelectorFunctor {
 public:
     bool operator()(const CSSSelector* selector)
     {
-        return selector->isUnknownPseudoElement() || selector->isCustomPseudoElement();
```
That would be a good opportunity to convert the whole functor class to a simple lambda.

annevk committed:
querySelector() throws exception for -webkit-prefixed pseudo elements
https://bugs.webkit.org/show_bug.cgi?id=149160
rdar://99299129

Reviewed by Antti Koivisto.

SelectorHasInvalidSelectorFunctor which is only used by querySelector()
and friends treated custom pseudo-elements as invalid. This is wrong.
querySelector("::placeholder") and querySelector("::-webkit-asdf") are
perfectly acceptable inputs, they are just supposed to return null.

Also do some minor code cleanup while in the general area.

Canonical link: https://commits.webkit.org/272337@main
